### DISCUSSION TITLE
【質問】演習　１文字表示NGの件

### DISCUSSION THREAD

**Discussion by tsukada-shigeki** [toppers/users-forum#69] [TOPPERS Seminar]
竹内様
お世話になります。
１）新基礎３編：１日目　LCD描画プログラムに関してお尋ねします。
２）p157　文字表示の解答に従ってDrawCharを完成しビルドできましたが、
　ABC表示NGとなります（添付写真参照願います。）
３）本日作成のlcd_charを添付致しますので原因をご教示願います。
以上宜しくお願い致します。
塚田
![ABC表示NG](https://user-images.githubusercontent.com/106538497/171978988-b0461db2-985f-4fca-8fcd-b72dd08748b7.jpg)
[演習１文字表示抜粋.pdf](https://github.com/toppers/users-forum/files/8836926/default.pdf)
![スクリーンショット (999)](https://user-images.githubusercontent.com/106538497/171979033-48a1e85d-dd0f-4704-ba1e-ed7977cfcd97.png)
[lcd_char1.zip](https://github.com/toppers/users-forum/files/8836927/lcd_char1.zip)

**Comment by Ryohsuke-Takeuchi** [comment]
塚田様
lcd_char1.cの問題ある行を修正したものをいかに添付します。
フォントは、縦pFont-Height、横pFont->Widthのイメージデータなので、
for文で、0からpFont-Height-1のカラムを繰り返しとりだし、
その繰り返しの中で、各カラムのイメージを0～pFont->Width-1を参照して、TextColorかBackColorを設定します。

static void
DrawChar(uint16_t x0, uint16_t y0, const uint8_t *c)
{
	LCD_Handler_t *hlcd = DrawProp.hlcd;
	sFONT    *pFont = DrawProp.pFont;
	
	uint32_t h,w,stride;
	uint8_t *pchar;
	
	stride = (pFont->Width +7)/8;
	lcd_setAddrWindow(hlcd,x0,y0,x0+pFont->Width-1,y0+pFont->Height-1);
	for(h =0; h< pFont->Height; h++){			/* フォントの0～pFont->Height-1まで繰り返す */
		pchar =((uint8_t *)c + stride * h);
#if 0	/* 修正します */
		for(w =0; w< pFont->Height; w++){
#else	/* 修正します */
		for(w =0; w< pFont->Width; w++){		/* フォントの各高さ位置で、0～pFont->Width-1まで繰り返す */
#endif	/* 修正します */
			if(pchar[w/8] &(1 <<(7-(w % 8)))){
				lcd_writedata2(hlcd,DrawProp.TextColor,1);
			}
			else{
				lcd_writedata2(hlcd,DrawProp.BackColor,1);
			}
		}
	}
}


通常の実装では、1pixelづつイメージを書き込むとオーバーヘッドが多いので
プロック化してデータを書き込みます。

**Comment by Ryohsuke-Takeuchi** [comment]
![lcd_char_modified](https://user-images.githubusercontent.com/91311953/172042934-ef98e883-df6d-47d4-b925-6ae1c34b2f85.png)

**Reply by tsukada-shigeki** [reply]
竹内様
お世話になります。
早速のご回答有難うございました。
只今試しましたところ問題なく動作致しました。有難うございました。
塚田
![ABC表示ok](https://user-images.githubusercontent.com/106538497/172056575-c5318f9c-1fb0-4cf0-b484-8855459f8ca8.jpg)
